src/compose.ts

```diff
@@ -346,3 +364,97 @@ function normalizeKeyValuePairs(
 		.fromPairs()
 		.value();
 }
+
+/**
+ * Parses a composition and returns a list of image descriptors that can
+ * be used to pull or build a service image. The given composition version
+ * must be equal to `DEFAULT_SCHEMA_VERSION`, or an exception is thrown.
+ * Normalise the composition before passing it to this function.
+ */
+async function expandContext(
+	c: Composition,
+	fileResolverCb: (path: string) => Promise<ReadStream>,
+): Promise<Composition> {
+	for (const key of Object.keys(c.services)) {
+		let envFilePaths = c.services[key].env_file;
+		if (envFilePaths !== undefined) {
+			if (!Array.isArray(envFilePaths)) {
+				envFilePaths = [envFilePaths];
+			}
+			for (const envFilePath of envFilePaths) {
+				filePathValidator(envFilePath);
+				c.services[key].environment = await readAndExpandEnvFile(
+					envFilePath,
+					fileResolverCb,
+					c.services[key].environment,
+				);
+			}
+		}
+	}
+	return c;
+}
+
+async function readAndExpandEnvFile(
+	envFile: string,
+	fileResolverCb: (path: string) => Promise<ReadStream>,
+	environment: Dict<string> = {},
+): Promise<Dict<string>> {
+	const readline = require('readline');
+	const { once } = require('events');
+	let envVarDict: Dict<string> = {};
+
+	let readStreamError;
+
+	// instantiate readStream from callback to add eventlistener to it
+	const readStream = await fileResolverCb(envFile);
+	const lineReader = readline.createInterface({
+		input: readStream,
+		crlfDelay: Infinity,
+	});
+
+	// get error from  stream reader and close linereader
+	// no race condition as the lineReader is paused until an event listener is registered
+	readStream.on('error', (readError) => {
+		readStreamError = readError;
+		lineReader.close();
+	});
+
+	// process each line on event
+	// now readStream will be evaluated
+	lineReader.on('line', (line: string) => {
+		envVarDict = { ...envVarDict, ...normalizeKeyValuePairs([line]) };
+	});
+
+	// wait until all lines read or stream error error occured
+	await once(lineReader, 'close');
+
+	if (readStreamError !== undefined) {
+		throw readStreamError;
+	}
+
+	for (const envKey of Object.keys(envVarDict)) {
+		if (environment === undefined) {
+			environment = { envKey: envVarDict[envKey] };
+		} else if (!(envKey in environment)) {
+			environment[envKey] = envVarDict[envKey];
+		} else {
+			// don't overwrite the existing env as the compose file has majority
```

Comment:
Should "majority" be "priority" in comment?

Reply:
Applied